feat: add image signing verification for ECR images signed by AWS Signer

ginglis13 opened the pull request:
<!--
Tips:
- Please read CONTRIBUTING.md to understand our process and our requests for PRs.
- Please file an issue before creating a PR so we can discuss the change and confirm it's not already being worked on.
-->

**Issue number:**

Related: https://github.com/bottlerocket-os/bottlerocket/issues/4684

**Description of changes:**

In containerd 2.0, support for [image verification plugins](https://github.com/containerd/containerd/blob/main/docs/image-verification.md) landed as part of the new [containerd image transfer service](https://github.com/containerd/containerd/blob/main/docs/transfer.md). This PR includes an image verifier binary and its required dependencies to verify ECR images signed via the [AWS Signer service](https://docs.aws.amazon.com/signer/latest/developerguide/Welcome.html).

New packages:
* [notation](https://github.com/notaryproject/notation): a CLI for signing and verifying artifacts.
* [ecr-credential-helper](https://github.com/awslabs/amazon-ecr-credential-helper): a Docker credential helper for authenticating with ECR
* [aws-signer-notation-plugin](https://github.com/aws/aws-signer-notation-plugin): a plugin for notation to interact with AWS signer. As packaged, it will write the AWS signer commercial and gov root CAs to a location expected by the notation CLI
* notation-image-verifier: a first party source that sets points notation to the config, library, and cache directories packaged via the core-kit, wraps the notation CLI and runs `notation verify` on the incoming image.


**Testing done:**

I built the core-kit with https://github.com/bottlerocket-os/bottlerocket-settings-sdk/pull/106 and changes to my local Bottlerocket aws-k8s-1.34 variant to include these new packages / settings / configuration-files. I then tested, in both `us-west-2` and `us-gov-west-1`:

1. Configure my node with `settings.image-verifier-plugins.notation.trustpolicy=""ewog..."` which was base64 encoded from a trustpolicy.json like:

```json
{
   "version":"1.0",
   "trustPolicies":[
      {
         "name":"aws-signer-tp",
         "registryScopes":[
            "*"
         ],
         "signatureVerification":{
            "level":"strict",
            "override": {
                "revocation": "skip"
            }
         },
         "trustStores":[
            "signingAuthority:aws-signer-ts",
            "signingAuthority:aws-us-gov-signer-ts"
         ],
         "trustedIdentities":[
            "arn:aws:signer:us-west-2:<redacted>:/signing-profiles/notation_test"
         ]
      }
   ]
}
```

2. ctr pull an unsigned image from a private registry in my account. Pull fails due to lack of signature.
3. ctr pull a signed image from a private registry in my account. Pull succeeds.
4. Repeat 2 and 3 with public registries, and all of this on aws-k8s-1.34-fips

Example of failure:

```
bash-5.1# ctr i pull docker.io/library/hello-world:latest
ctr: image verifier bindir blocked pull of docker.io/library/hello-world:latest with digest sha256:56433a6be3fda188089fb548eae3d91df3ed0d6589f7c2656121b911198df065 for reason: verifier notation-image-verifier rejected image (exit code 1): verifying image: docker.io/library/hello-world:latest
image verification failed: log: failed to create directory: mkdir /root/.ecr: read-only file systemWarning: Always verify the artifact using digest(@sha256:...) rather than a tag(:latest) because resolved digest may not point to the same signed artifact, as tags are mutable.
Error: signature verification failed: no signature is associated with "docker.io/library/hello-world@sha256:56433a6be3fda188089fb548eae3d91df3ed0d6589f7c2656121b911198df065", make sure the artifact was signed successfully
```

Pull succeeding has no additional output to normal pull without image verification.

Processes also are inheriting the `runtime_t` selinux label:

```
bash-5.1# ps -eZ | grep notation
system_u:system_r:runtime_t:s0     2922 ?        00:00:00 notation-image-
system_u:system_r:runtime_t:s0     2928 ?        00:00:00 notation
bash-5.1# ps -eZ | grep containerd
system_u:system_r:runtime_t:s0     1574 ?        00:00:07 containerd
system_u:system_r:runtime_t:s0     1626 ?        00:00:01 containerd
system_u:system_r:runtime_t:s0     1936 ?        00:00:00 containerd-shim
system_u:system_r:runtime_t:s0     2414 ?        00:00:00 containerd-shim
```


**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

ginglis13 committed:
packages: add notation

Package the Notation CLI, a tool for signing and verifying artifacts.
Include template for rendering a notation trustpolicy document provided
via Bottlerocket settings.

Signed-off-by: Gavin Inglis <giinglis@amazon.com>

ginglis13 committed:
packages: add ecr-credential-helper

Add the ECR credential helper for use by notation CLI when
authenticating to ECR registries. The change includes a bare-bones
docker config written to /root/.docker/config.json, as the notation CLI
will find credential helpers via this configuration file. Also configure
mounts the ECR credentials cache directory.

Signed-off-by: Gavin Inglis <giinglis@amazon.com>

ginglis13 committed:
packages: add aws-signer-notation-plugin

Add the AWS Signer plugin for notation. Place the plugin in the plugins
directory packaged by notation so that it is discoverable without the
need to import via the notation tool. Provide the AWS Signer root CAs
for the same reason.

Signed-off-by: Gavin Inglis <giinglis@amazon.com>

ginglis13 committed:
sources: add notation-image-verifier

Add a simple wrapper of the notation CLI that adheres to containerd's
verification plugin interface. The binary configures notation's config,
cache, library directories to what the core-kit packages and simply runs
"notation verify" against the input image.

Credentials are handled by the ecr-credential-helper, which notation
discovers via a Docker config file at /root/.docker/config.json.

Signed-off-by: Gavin Inglis <giinglis@amazon.com>

ginglis13 commented on packages/aws-signer-notation-plugin/aws-signer-notation-root.crt:
```diff
@@ -0,0 +1,15 @@
+-----BEGIN CERTIFICATE-----
```
these certs are linked in this guide: https://docs.aws.amazon.com/signer/latest/developerguide/image-signing-prerequisites.html

arnaldo2792 replied:
Is there a way to know when we have to rotate the certificates? Maybe a script or something we can run?

ginglis13 commented on sources/notation-image-verifier/main.go:
```diff
@@ -0,0 +1,46 @@
+
+	fmt.Printf("verifying image: %s\n", *name)
+
+	// Override the default notation paths to what's packaged in packages/notation
```
I'll update this with a direct link to https://notaryproject.dev/docs/user-guides/how-to/directory-structure/ in the comment

ginglis13 commented on packages/notation/notation.spec:
```diff
@@ -0,0 +1,92 @@
+%{_cross_tmpfilesdir}/notation.conf
+%dir %{_cross_factorydir}%{_cross_sysconfdir}/notation
+%dir %{_cross_factorydir}%{_cross_sysconfdir}/notation/plugins
+%dir %{_cross_factorydir}%{_cross_sysconfdir}/notation/truststore/x509/signingAuthority
```
See https://github.com/notaryproject/specifications/blob/main/specs/trust-store-trust-policy.md#trust-store

ginglis13 commented on packages/aws-signer-notation-plugin/aws-signer-notation-plugin.spec:
```diff
@@ -0,0 +1,98 @@
+%dir %{_cross_factorydir}%{_cross_sysconfdir}/notation/truststore/x509/signingAuthority/aws-signer-ts/
+%dir %{_cross_factorydir}%{_cross_sysconfdir}/notation/truststore/x509/signingAuthority/aws-us-gov-signer-ts/
+%{_cross_factorydir}%{_cross_sysconfdir}/notation/truststore/x509/signingAuthority/aws-signer-ts/aws-signer-notation-root.crt
+%{_cross_factorydir}%{_cross_sysconfdir}/notation/truststore/x509/signingAuthority/aws-us-gov-signer-ts/aws-us-gov-signer-notation-root.crt
```
See the notation spec https://github.com/notaryproject/specifications/blob/main/specs/trust-store-trust-policy.md#trust-store

ginglis13 commented:
^ force push adds
```
C /etc/notation - - - - /usr/share/factory/etc/notation
```

to the notation tmpfiles to copy over the factory files.

arnaldo2792 commented on packages/notation/notation.spec:
```diff
@@ -0,0 +1,92 @@
+Source2: notation-trust-policy-json
+Source3: notation-tmpfiles.conf
+
+BuildRequires: git
```
You don't need git for this build, right? No `git` repos are initialized, and there aren't patches applied.

arnaldo2792 commented on packages/notation/notation-tmpfiles.conf:
```diff
@@ -0,0 +1,2 @@
+d /var/cache/notation - - - -
+C /etc/notation - - - - /usr/share/factory/etc/notation
```
If the source factory is the exact same path as the target path, you can avoid it.

arnaldo2792 commented on packages/ecr-credential-helper/ecr-credential-helper.spec:
```diff
@@ -0,0 +1,81 @@
+# cross_go_configure cd's to the correct GOPATH location
+%cross_go_configure %{goimport}
+
+export GO_MAJOR="1.24"
```
Why do you need this? isn't `GO_MAJOR` already `1.24`?

arnaldo2792 commented on packages/aws-signer-notation-plugin/aws-signer-notation-plugin.spec:
```diff
@@ -0,0 +1,98 @@
+# Place the binaries where notation expects them.
+install -d %{buildroot}%{_cross_bindir}/notation-plugins
+install -d %{buildroot}%{_cross_bindir}/notation-plugins/plugins
+install -d %{buildroot}%{_cross_bindir}/notation-plugins/plugins/com.amazonaws.signer.notation.plugin
```
You could create a `global` to reduce some of the verbosity of the path.

arnaldo2792 commented on packages/aws-signer-notation-plugin/aws-signer-notation-plugin.spec:
```diff
@@ -0,0 +1,98 @@
+install -p -m 0755 fips/notation-com.amazonaws.signer.notation.plugin %{buildroot}%{_cross_fips_bindir}/notation-plugins/plugins/com.amazonaws.signer.notation.plugin/notation-com.amazonaws.signer.notation.plugin
+
+# Add the notation config truststore directories.
+install -d %{buildroot}%{_cross_factorydir}%{_cross_sysconfdir}/notation/truststore/x509/signingAuthority/aws-signer-ts/
```
Same for this, a `global` could help with the verbosity.

arnaldo2792 commented on packages/notation-image-verifier/notation-image-verifier.spec:
```diff
@@ -0,0 +1,35 @@
+Name: %{_cross_os}%{workspace_name}
+Version: 0.1.0
+Release: 1%{?dist}
+Summary: Binary for containerd image verification plugin
```
```suggestion
Summary: A notation-based containerd image verification plugin
```

arnaldo2792 commented on packages/notation-image-verifier/notation-image-verifier.spec:
```diff
@@ -0,0 +1,35 @@
+cp -r %{_builddir}/sources/%{workspace_name}/* .
+
+%build
+export GO_MAJOR="1.24"
```
Same as elsewhere, `GO_MAJOR` is already `1.24`. Since we control the sources of this package, you could just use 1.25 right away.

arnaldo2792 commented on packages/notation-image-verifier/notation-image-verifier.spec:
```diff
@@ -0,0 +1,35 @@
+%install
+install -d %{buildroot}%{_cross_bindir}/image-verifiers
+install -d %{buildroot}%{_cross_bindir}/image-verifiers/bin
+install -p -m 0755 notation-image-verifier %{buildroot}%{_cross_bindir}/image-verifiers/bin
```
You still need to add the `cross_attribution` macro here.

ginglis13 replied:
Since `notation-image-verifier` doesn't pull in any 3rd party modules, there are no dependencies to vendor. the `cross_scan_attribution` macro need a vendor directory to process. We could create an empty one and produce an empty attribution, but I'm just going to omit that instead.

arnaldo2792 commented on sources/Cargo.toml:
```diff
@@ -91,6 +91,7 @@ datastore = { version = "0.1", path = "api/datastore" }
 dogtag = { version = "0.1", path = "dogtag" }
 early-boot-config-provider = { version = "0.1", path = "early-boot-config/early-boot-config-provider" }
 generate-readme = { version = "0.1", path = "generate-readme" }
+image-verifier = { version = "0.1", path = "image-verifier"  }
```
You should only add this here if the package was a Rust program

```suggestion
```

arnaldo2792 commented on packages/os/os.spec:
```diff
@@ -82,6 +82,7 @@ Requires: %{_cross_os}corndog
 Requires: %{_cross_os}certdog
 Requires: %{_cross_os}driverdog
 Requires: %{_cross_os}ghostdog
+Requires: %{_cross_os}notation-image-verifier
```
You may not want this here, if anything it should go in `release`. But I'll say make this an opt-in and we will include it in the variant definition.

bcressey commented on packages/notation/notation-trust-policy-json:
```diff
@@ -0,0 +1,7 @@
+image-verifier-plugins = "v1"
+std = { version = "v1", helpers = ["base64_decode"] }
++++
+{{~#if settings.image-verifier-plugins.notation.trustpolicy~}}
+{{base64_decode settings.image-verifier-plugins.notation.trustpolicy}}
+{{~/if~}}
```
Should we emit a `{}` empty policy if we don't have a trust policy, or there's no need?

ginglis13 replied:
`notation` will fail with an empty policy. IMO there's no need - we're already making it a requirement to provide a valid trustpolicy to use the notation image verifier.

The alternative would be we provide a default trustpolicy like

```
{
  "version": "1.0",
  "trustPolicies": [
    {
      "name": "default",
      "registryScopes": ["*"],
      "signatureVerification": {
        "level": "skip"
      }
    }
  ]
}
```

that would essentially be a no-op for signature verification, WDYT?

bcressey commented on packages/notation/notation.spec:
```diff
@@ -0,0 +1,92 @@
+Name: %{_cross_os}%{gorepo}
+Version: %{rpmver}
+Release: 1%{?dist}
+Summary:  A CLI tool to sign and verify artifacts.
```
nit:

```suggestion
Summary: A CLI tool to sign and verify artifacts.
```

bcressey commented on packages/notation/notation.spec:
```diff
@@ -0,0 +1,92 @@
+%setup -T -D -n %{gorepo}-%{gover} -b 1 -q
+
+%build
+
```
nit:
```suggestion
```

arnaldo2792 commented on sources/notation-image-verifier/main.go:
```diff
@@ -0,0 +1,46 @@
+	// Bottlerocket doesn't set a $HOME.
+	os.Setenv("HOME", "/root")
+
+	cmd := exec.Command("notation", "verify", *name)
```
Is there a problem if you use the full path to `notation`? At least that's what we do in rust programs when we call other helper binaries.

ginglis13 replied:
good call, I'm going to update to the absolute path

I'll keep this unresolved as I keep getting through the lower hanging fruit feedback - I'll have to consider the FIPS case for that path.

ginglis13 replied:
NVM - I'll point to `/usr/bin/notation`, we have an overlay that already handles this :D https://github.com/bottlerocket-os/bottlerocket-core-kit/blob/39a939781bd126aeaf3ef7e3b3185f6c18a8b439/packages/release/usr-bin.mount.in#L11

arnaldo2792 commented on packages/aws-signer-notation-plugin/aws-signer-notation-plugin.spec:
```diff
@@ -0,0 +1,98 @@
+
+# Place the binaries where notation expects them.
+install -d %{buildroot}%{_cross_bindir}/notation-plugins
+install -d %{buildroot}%{_cross_bindir}/notation-plugins/plugins
```
I saw you have control over this path, it will be best to follow fedora's guidelines and install them in `libexec`.

arnaldo2792 commented on sources/notation-image-verifier/main.go:
```diff
@@ -0,0 +1,46 @@
+
+	if *digest == "" || *name == "" {
+		// containerd only captures the stdout of image verification binaries, hence no
+		// write to stderr here (and elsewhere in the program).
```
So are you saying `containerd` only cares about what's written to the `stdout`? does it expect any format in the output or does it capture `stdout` for logging purposes?

ginglis13 replied:
It captures stdout for logging - stdout is where it expects the "reason" for image verification rejection: https://github.com/containerd/containerd/blob/main/docs/image-verification.md#image-pull-judgement

bcressey commented on packages/ecr-credential-helper/ecr-credential-helper.spec:
```diff
@@ -0,0 +1,81 @@
+%prep
+%autosetup -n %{gorepo}-%{gover} -p1
+%cross_go_setup %{gorepo}-%{gover} %{goproject} %{goimport}
+
```
```suggestion
```

arnaldo2792 commented on sources/notation-image-verifier/main.go:
```diff
@@ -0,0 +1,46 @@
+	// Override the default notation paths to what's packaged in packages/notation
+	os.Setenv("NOTATION_CONFIG", "/etc/notation")
+	os.Setenv("NOTATION_CACHE", "/var/cache/notation")
+	os.Setenv("NOTATION_LIBEXEC", "/usr/bin/notation-plugins")
```
As mentioned elsewhere, it will be best to refer to `libexec`

arnaldo2792 commented on sources/notation-image-verifier/main.go:
```diff
@@ -0,0 +1,46 @@
+
+	// The notation CLI looks in $HOME/.docker/config.json for credentials.
+	// Bottlerocket doesn't set a $HOME.
+	os.Setenv("HOME", "/root")
```
TIL!

arnaldo2792 commented on Cargo.toml:
```diff
@@ -101,6 +101,7 @@ members = [
   "packages/netdog",
```
nit: for consistency, you could separate the `source` and `package` commits.

arnaldo2792 commented on packages/containerd-2.1/image-verification-toml:
```diff
@@ -0,0 +1,11 @@
+{{~#if settings.image-verifier-plugins.notation.trustpolicy~}}
+[plugins]
+  [plugins."io.containerd.image-verifier.v1.bindir"]
+    bin_dir = "/usr/bin/image-verifiers/bin"
```
If we control this, I would say `libexec` is the right place for the plugins

bcressey commented on packages/aws-signer-notation-plugin/Cargo.toml:
```diff
@@ -0,0 +1,17 @@
+[lib]
+path = "../packages.rs"
+
+[[package.metadata.build-package.external-files]]
```
nit: missing `releases-url` section

bcressey commented on sources/notation-image-verifier/main.go:
```diff
@@ -0,0 +1,46 @@
+	if err != nil {
+		fmt.Printf("image verification failed: %s\n", string(output))
+		os.Exit(1)
+	}
```
Is there a concern that we're not validating the digest also? Should we try some approach where we append `@<digest>` to `name` if it doesn't already end in the digest? Or parse the output to see if it indicates the same digest?

ginglis13 replied:
I'll update this to instead construct the image reference we pass to `notation` like `<uri>:<digest>`. the `notation` CLI does warn against using named tags, since tags can be immutable. I'll have to parse the incoming `name` from containerd a bit to construct this.

I'll also pull some of this logic out of `main` and add some unit testing.

bcressey commented on packages/notation/notation-tmpfiles.conf:
```diff
@@ -0,0 +1,2 @@
+d /var/cache/notation - - - -
```
What ends up in the cache? If it's not vital, we can clean it up with an `R` directive on each boot before creating it. One less bit of state to worry about across upgrades and downgrades.

ginglis13 replied:
The cache has Certificate Revocation List hashes: https://notaryproject.dev/docs/user-guides/how-to/directory-structure/#structure

I'll add the `R` directive 👍

bcressey commented on packages/notation-image-verifier/notation-image-verifier.spec:
```diff
@@ -0,0 +1,35 @@
+Summary: Binary for containerd image verification plugin
+License: Apache-2.0 OR MIT
+URL: https://github.com/bottlerocket-os/bottlerocket
+BuildRequires: %{_cross_os}glibc-devel
```
This package should probably require the notation CLI and perhaps the signer plugin also.

bcressey commented on packages/os/os.spec:
```diff
@@ -83,6 +82,7 @@ Requires: %{_cross_os}corndog
 Requires: %{_cross_os}certdog
 Requires: %{_cross_os}driverdog
 Requires: %{_cross_os}ghostdog
+Requires: %{_cross_os}notation-image-verifier
```
I'd probably move this Requires to containerd-2.1:
1. It's the only one configured to use it.
2. We're a bit short on space for older variants.

ginglis13 commented on packages/ecr-credential-helper/ecr-credential-helper.spec:
```diff
@@ -0,0 +1,81 @@
+%{_cross_attribution_file}
+%{_cross_attribution_vendor_dir}
+%dir /root/.docker
+/root/.docker/config.json
```
Working through the initial feedback first, but I'm going to look deeper into notation's config.json - might be a more appropriate place to do this https://notaryproject.dev/docs/user-guides/how-to/directory-structure/#general-configuration

ginglis13 commented:
^ force push addresses feedback, the more notable changes being:

* move the aws-signer-notation-plugin and notation-image-verifier binaries to libexecdir
* notation-image-verifier now constructs an image reference from the incoming image sha to pass to notation

ginglis13 commented:
I've rerun manual tests on a custom `aws-k8s-1.34` variant along with the latest in https://github.com/bottlerocket-os/bottlerocket-settings-sdk/pull/106. 

User Data:
```
[settings.image-verifier-plugins.notation]
trustpolicy = "ewogIC..."

[settings.image-verifier-plugins]
enabled = true
```

Successfully pulled my image in private registry, failed to pull `docker.io/library/hello-world:latest` as expected

bcressey commented on packages/containerd-2.1/containerd-2.1.spec:
```diff
@@ -38,13 +39,13 @@ Source1000: clarify.toml
 Requires: %{_cross_os}runc
 Requires: %{name}(optimized-gunzip)
 Requires: %{name}(binaries)
+Requires: %{_cross_os}notation-image-verifier
```
If the only settings plugin that will include these new settings is `aws-ecs-3`, then we should omit this and just add it at the variant level. If you're planning to include everywhere containerd 2.1 is included, then this is fine.

Maybe these new files should go into the `notation-image-verifier` package? It doesn't seem like `containerd-2.1` has to do anything special.

ginglis13 replied:
It's safest for now to pull this into the notation-image-verifier and include that in the aws-ecs-3 variant.

bcressey commented on packages/notation-image-verifier/notation-image-verifier.spec:
```diff
@@ -0,0 +1,66 @@
+
+%install
+install -d %{buildroot}%{_cross_libexecdir}/image-verifiers
+install -d %{buildroot}%{_cross_libexecdir}/image-verifiers/bin
```
nit: can condense into one line:

```suggestion
install -d %{buildroot}%{_cross_libexecdir}/image-verifiers/bin
```

(same with fips)

bcressey commented on packages/notation/notation-tmpfiles.conf:
```diff
@@ -0,0 +1,2 @@
+R /var/cache/notation - - - -
```
Does this directory get automatically created?

ginglis13 replied:
it gets created at runtime by notation: https://github.com/notaryproject/notation/blob/e49f07cd19d47fb5c79a0c56321c70bf13d525ef/internal/revocation/revocation.go#L34

bcressey commented on sources/notation-image-verifier/main.go:
```diff
@@ -0,0 +1,73 @@
+	"strings"
+)
+
+// constructImageReference creates a notation-compatible image reference
+// by combining the URI (name without tag) with the digest.
+// Example: "docker.io/library/hello-world:latest" + "sha256:abc123"
+//
+//	-> "docker.io/library/hello-world@sha256:abc123"
+func constructImageReference(name, digest string) string {
+	// Remove tag if present (everything after the last colon that's not part of a port)
+	uri := name
+	if lastColon := strings.LastIndex(name, ":"); lastColon != -1 {
+		// Check if this colon is part of a hostname:port (contains slash after colon)
+		if !strings.Contains(name[lastColon:], "/") {
+			uri = name[:lastColon]
+		}
+	}
+	return uri + "@" + digest
```
With one of my own images:
```
# Ideally we are checking a signature for this digest.
❯ crane digest 760038019737.dkr.ecr.us-west-2.amazonaws.com/bottlerocket-sdk:v0.665.0
sha256:4af3f570d6d60ca101f0aa05b927377191ac256a2397759cc64a64c8db325e48

# And indeed that seems to exist:
❯ crane digest 760038019737.dkr.ecr.us-west-2.amazonaws.com/bottlerocket-sdk:v0.665.0@sha256:4af3f570d6d60ca101f0aa05b927377191ac256a2397759cc64a64c8db325e48
sha256:4af3f570d6d60ca101f0aa05b927377191ac256a2397759cc64a64c8db325e

# But I have an older image in the same repository:
❯ crane digest 760038019737.dkr.ecr.us-west-2.amazonaws.com/bottlerocket-sdk:v0.60.0
sha256:3fb0141c578f86d3ae5143f7d21bed41874045dfcf6cd6360e1331faf1c35896

# And I can use that ref to refer to the older image:
❯ crane digest 760038019737.dkr.ecr.us-west-2.amazonaws.com/bottlerocket-sdk:v0.665.0@sha256:3fb0141c578f86d3ae5143f7d21bed41874045dfcf6cd6360e1331faf1c35896
sha256:3fb0141c578f86d3ae5143f7d21bed41874045dfcf6cd6360e1331faf1c35896
```

Is Notation better at handling this? I worry that it'd fall back to checking whether any image in the registry had a signature for the digest, instead of the tag.

If the only thing that's safe is to accept images without a tag and with the digest value at the end, then we should enforce that.

bcressey replied:
Chatted with @ginglis13 offline. If we trust that containerd has already resolved the ref to a digest, then the alternate forms of the image ref don't really matter. We just have to construct a ref that refers unambiguously to the provided digest, then look for a matching signature.

ginglis13 replied:
Some manual testing and looking at containerd sources confirms that containerd will resolve the digest from the incoming ref first before invoking verifier binaries. The notation-image-verifier only uses the incoming resolved digest to verify signatures.

ginglis13 commented:
^ last 2 force pushes handle the case of `registry/repo:tag@digest` as the input image name to the `notation-image-verifier` and package the containerd drop-in via the `notation-image-verifier` package. I've also dropped the changed to `containerd-2.1`, since those changes were just to pull in the verifier binary as a direct dependency and package the template file for the verification plugins containerd config drop-in.

ginglis13 commented:
^ force push packages a `notation` config.json at `/etc/notation`, configuring ecr-credential-helper to be used. Also moves the notation package to depend on ecr-credential-helper.

bcressey commented on packages/notation/notation.spec:
```diff
@@ -0,0 +1,94 @@
+Requires: %{name}(binaries)
+Requires: %{_cross_os}ecr-credential-helper
+
+Provides: %{_cross_os}notation
```
nit: can cut this, it should be provided by the package's name:

```suggestion
```

ginglis13 committed:
filesystem: add /root/.docker and /root/.ecr

Signed-off-by: Gavin Inglis <giinglis@amazon.com>

ginglis13 commented:
^ force push:
* reverts back to using `/root/.docker` for configuring `"credsStore": "ecr-login"` - helpful when using the docker CLI directly on an ecs- variant.
* adds `secret_t` mounts for `/root/.docker` for the docker config and `/root/.ecr` for the ecr-credential-helper cache
